File: C++/matrix.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <numeric>

using namespace std;
using namespace chrono;

constexpr int N = 2000;
constexpr int ITERATIONS = 5;

void matmul(vector<double>& C, const vector<double>& A, const vector<double>& B, int n) {
    for (int i = 0; i < n; i++) {
        for (int k = 0; k < n; k++) {
            double aik = A[i * n + k];
            for (int j = 0; j < n; j++) {
                C[i * n + j] += aik * B[k * n + j];
            }
        }
    }
}

void transpose(vector<double>& T, const vector<double>& M, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            T[j * n + i] = M[i * n + j];
        }
    }
}

void add_matrices(vector<double>& C, const vector<double>& A, const vector<double>& B, int n) {
    for (int i = 0; i < n * n; i++) {
        C[i] = A[i] + B[i];
    }
}

int main() {
    vector<double> A(N * N), B(N * N), C(N * N), T(N * N);
    mt19937 rng(42);
    uniform_real_distribution<double> dist(0.0, 1.0);
    for (auto& x : A) x = dist(rng);
    for (auto& x : B) x = dist(rng);
    
    cout << "C++ Matrix Benchmark (" << N << "x" << N << ", " << ITERATIONS << " iterations)\n";
    
    double total_ms = 0;
    double checksum = 0;
    
    for (int iter = 0; iter < ITERATIONS; iter++) {
        fill(C.begin(), C.end(), 0.0);
        
        auto t0 = high_resolution_clock::now();
        matmul(C, A, B, N);
        auto t1 = high_resolution_clock::now();
        
        total_ms += duration<double, milli>(t1 - t0).count();
        for (const auto& x : C) checksum += x;
    }
    
    double avg_ms = total_ms / ITERATIONS;
    double ops = (double)N * N * N * 2;
    double gflops = (ops * ITERATIONS) / (total_ms / 1000.0) / 1e9;
    
    cout << "Multiply: " << avg_ms << " ms (" << gflops << " GFLOPS)\n";
    
    volatile double sink = 0;
    auto t0 = high_resolution_clock::now();
    for (int iter = 0; iter < ITERATIONS; iter++) {
        transpose(T, A, N);
        for (const auto& x : T) checksum += x;
        sink += checksum;
    }
    auto t1 = high_resolution_clock::now();
    avg_ms = duration<double, milli>(t1 - t0).count() / ITERATIONS;
    cout << "Transpose: " << avg_ms << " ms\n";
    
    t0 = high_resolution_clock::now();
    for (int iter = 0; iter < ITERATIONS; iter++) {
        add_matrices(C, A, B, N);
    }
    t1 = high_resolution_clock::now();
    avg_ms = duration<double, milli>(t1 - t0).count() / ITERATIONS;
    cout << "Add: " << avg_ms << " ms\n";
    
    return 0;
}

File: Rust/matrix.rs
use rand::Rng;

fn matmul(C: &mut [f64], A: &[f64], B: &[f64], n: usize) {
    for i in 0..n {
        for k in 0..n {
            let aik = A[i * n + k];
            for j in 0..n {
                C[i * n + j] += aik * B[k * n + j];
            }
        }
    }
}

fn transpose(T: &mut [f64], M: &[f64], n: usize) {
    for i in 0..n {
        for j in 0..n {
            T[j * n + i] = M[i * n + j];
        }
    }
}

fn add_matrices(C: &mut [f64], A: &[f64], B: &[f64], _n: usize) {
    for i in 0..C.len() {
        C[i] = A[i] + B[i];
    }
}

fn main() {
    const N: usize = 2000;
    const ITERATIONS: usize = 5;
    
    println!("Rust Matrix Benchmark ({}x{}, {} iterations)", N, N, ITERATIONS);
    
    let mut rng = rand::thread_rng();
    let A: Vec<f64> = (0..N * N).map(|_| rng.gen()).collect();
    let B: Vec<f64> = (0..N * N).map(|_| rng.gen()).collect();
    let mut C: Vec<f64> = vec![0.0; N * N];
    let mut T: Vec<f64> = vec![0.0; N * N];
    
    let mut total_ns: u64 = 0;
    let mut sum: f64 = 0.0;
    
    for _ in 0..ITERATIONS {
        C.fill(0.0);
        let t0 = std::time::Instant::now();
        matmul(&mut C, &A, &B, N);
        let t1 = std::time::Instant::now();
        total_ns += t1.duration_since(t0).as_nanos() as u64;
        sum += C.iter().sum::<f64>();
    }
    
    let avg_ms = (total_ns / ITERATIONS as u64) as f64 / 1e6;
    let ops = N as f64 * N as f64 * N as f64 * 2.0;
    let gflops = (ops * ITERATIONS as f64) / (total_ns as f64 / 1e9) / 1e9;
    
    println!("Multiply: {:.2} ms ({:.2} GFLOPS)", avg_ms, gflops);
    
    let t0 = std::time::Instant::now();
    for _ in 0..ITERATIONS {
        transpose(&mut T, &A, N);
        let sum = T.iter().sum::<f64>();
        std::hint::black_box(sum);
    }
    let t1 = std::time::Instant::now();
    let avg_ms = (t1.duration_since(t0).as_nanos() as f64 / ITERATIONS as f64) / 1e6;
    println!("Transpose: {:.2} ms", avg_ms);
    
    let t0 = std::time::Instant::now();
    for _ in 0..ITERATIONS {
        add_matrices(&mut C, &A, &B, N);
    }
    let t1 = std::time::Instant::now();
    let avg_ms = (t1.duration_since(t0).as_nanos() as f64 / ITERATIONS as f64) / 1e6;
    println!("Add: {:.2} ms", avg_ms);
}
